File: org.kevoree.modeling.microframework/src/main/java/org/kevoree/modeling/api/abs/AbstractKDimension.java
package org.kevoree.modeling.api.abs;

import org.kevoree.modeling.api.Callback;
import org.kevoree.modeling.api.KDimension;
import org.kevoree.modeling.api.KUniverse;
import org.kevoree.modeling.api.KView;
import org.kevoree.modeling.api.event.ListenerScope;
import org.kevoree.modeling.api.time.TimeTree;
import org.kevoree.modeling.api.ModelListener;

import java.util.HashMap;
import java.util.Map;

/**
 * Created by duke on 10/10/14.
 */
public abstract class AbstractKDimension<A extends KView, B extends KDimension, C extends KUniverse> implements KDimension<A, B, C> {

    private KUniverse _universe;

    private long _key;

    protected AbstractKDimension(KUniverse p_universe, long p_key) {
        this._universe = p_universe;
        this._key = p_key;
    }

    @Override
    public long key() {
        return _key;
    }

    @Override
    public C universe() {
        return (C) _universe;
    }

    @Override
    public void save(Callback<Throwable> callback) {
        universe().storage().save(this, callback);
    }

    @Override
    public void saveUnload(Callback<Throwable> callback) {
        universe().storage().saveUnload(this, callback);
    }

    @Override
    public void delete(Callback<Throwable> callback) {
        universe().storage().delete(this, callback);
    }

    @Override
    public void discard(Callback<Throwable> callback) {
        universe().storage().discard(this, callback);
    }

    @Override
    public void parent(Callback<B> callback) {
        //TODO
    }

    @Override
    public void children(Callback<B[]> callback) {
        //TODO
    }

    @Override
    public void fork(Callback<B> callback) {
        //TODO
    }

    @Override
    public synchronized A time(Long timePoint) {
        return internal_create(timePoint);
    }

    public void listen(ModelListener listener, ListenerScope scope) {
        universe().storage().eventBroker().registerListener(this, listener, scope);
    }

    protected abstract A internal_create(Long timePoint);

    @Override
    public boolean equals(Object obj) {
        if (!(obj instanceof AbstractKDimension)) {
            return false;
        } else {
            AbstractKDimension casted = (AbstractKDimension) obj;
            return casted._key == _key;
        }
    }

}

File: org.kevoree.modeling.microframework.typescript/src/main/ts/org/kevoree/modeling/api/abs/AbstractKDimension.ts
///<reference path="../Callback.ts"/>
///<reference path="../KDimension.ts"/>
///<reference path="../KUniverse.ts"/>
///<reference path="../KView.ts"/>
///<reference path="../time/TimeTree.ts"/>
///<reference path="../ModelListener.ts"/>
///<reference path="../../../../../java/util/HashMap.ts"/>
///<reference path="../../../../../java/util/Map.ts"/>
///<reference path="../../../../../java/util/Set.ts"/>

class AbstractKDimension<A, B, C> implements KDimension<A, B, C> {

  private _universe: KUniverse<any> = null;
  private _key: number = 0;
  private _timesCache: Map<number, A> = new HashMap<number, A>();

  constructor(p_universe: KUniverse, p_key: number) {
    this._universe = p_universe;
    this._key = p_key;
  }

  public key(): number {
    return this._key;
  }

  public universe(): C {
    return <C>this._universe;
  }

  public save(callback: Callback<Throwable>): void {
    this.universe().storage().save(this, callback);
  }

  public saveUnload(callback: Callback<Throwable>): void {
    this.universe().storage().saveUnload(this, callback);
  }

  public delete(callback: Callback<Throwable>): void {
    this.universe().storage().delete(this, callback);
  }

  public discard(callback: Callback<Throwable>): void {
    this.universe().storage().discard(this, callback);
  }

  public timeTrees(keys: number[], callback: Callback<TimeTree[]>): void {
    this.universe().storage().timeTrees(this, keys, callback);
  }

  public timeTree(key: number, callback: Callback<TimeTree>): void {
    this.universe().storage().timeTree(this, key, callback);
  }

  public parent(callback: Callback<B>): void {
  }

  public children(callback: Callback<Set<B>>): void {
  }

  public fork(callback: Callback<B>): void {
  }

  public time(timePoint: number): A {
    if (this._timesCache.containsKey(timePoint)) {
      return this._timesCache.get(timePoint);
    } else {
      var newCreatedTime: A = this.internal_create(timePoint);
      this._timesCache.put(timePoint, newCreatedTime);
      return newCreatedTime;
    }
  }

  public listen(listener: ModelListener): void {
    this.universe().storage().registerListener(this, listener);
  }

  public internal_create(timePoint: number): A {
    throw "Abstract method";
  }

}

